lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+        Fail()
+      end
+      return _to_error_state(source, "Unable to decode message")
+    end
+    _continue_perhaps(source)
+
+  fun ref _run_and_subsequent_activity(latest_metrics_id: U16,
+    ingest_ts: U64,
+    pipeline_time_spent: U64,
+    key_string: String val,
+    source: ConnectorSource[In] ref,
+    decoded: (In val| None val),
+    s: _StreamState,
+    message_id: (cwm.MessageId|None),
+    flags: cwm.Flags): Bool
+   =>
+    let decode_end_ts = WallClock.nanoseconds()
+    _metrics_reporter.step_metric(_pipeline_name,
+      "Decode Time in Connector Source", latest_metrics_id, ingest_ts,
+      decode_end_ts)
+    let latest_metrics_id' = latest_metrics_id + 1
+
+    let msg_uid = _msg_id_gen()
+
+    // TODO: We need a way to determine the key based on the policy
+    // for any particular connector. For example, the Kafka connector
+    // needs a way to provide the Kafka key here.
+
+    let initial_key =
+      if key_string isnt None then
+        key_string
+      else
+        // wat
+        // TODO [post-source-migration] should this be stream_id for messages that
+        // do not provide a key?
+        // What is a sane default here?
+        // What does it mean to use a unique key for each message?
+        msg_uid.string()
       end
 
-      source.expect(_header_size)
-      _header = true
+    // TOOD: We need a way to assign watermarks based on the policy
+    // for any particular connector.
+    if ingest_ts > _watermark_ts then
+      _watermark_ts = ingest_ts
+    end
 
-      ifdef linux then
-        true
-      else
-        false
+    (let is_finished, let last_ts) =
+      match decoded
+      | None =>
+        (true, ingest_ts)
+      | let d: In =>
+        _runner.run[In](_pipeline_name, pipeline_time_spent, d,
+          consume initial_key, ingest_ts, _watermark_ts, source_id,
+          source, _router, msg_uid, None, decode_end_ts,
+          latest_metrics_id, ingest_ts, _metrics_reporter)
+      end
+
+    match message_id
+    | let m_id: PointOfReference =>
+      if not (cwm.Ephemeral.is_set(flags) or
+        cwm.UnstableReference.is_set(flags)) then
+        s.last_seen = m_id
+      end
+    end
+
+    if is_finished then
+      let end_ts = WallClock.nanoseconds()
+      let time_spent = end_ts - ingest_ts
+
+      ifdef "detailed-metrics" then
+        _metrics_reporter.step_metric(_pipeline_name,
+          "Before end at Connector Source", 9999,
+          last_ts, end_ts)
       end
+
+      _metrics_reporter.pipeline_metric(_pipeline_name, time_spent +
+        pipeline_time_spent)
+      _metrics_reporter.worker_metric(_pipeline_name, time_spent)
+    end
+
+    _continue_perhaps(source)
+
+  fun ref _continue_perhaps(source: ConnectorSource[In] ref): Bool =>
+    ifdef "trace" then
+      @printf[I32]("NH: _continue_perhaps: %s\n".cstring(),
+        _header_size.string().cstring())
     end
+    source.expect(_header_size)
+    _header = true
+    _continue_perhaps2()
```

Comment:
Why is this a separate method?

Reply:
It is called twice.  {shrug}  The `ifdef` has distant origin in the original source & boundary code, preserving it seemed worthwhile.